File: bruteforce/KimJunYoung/ProgrammersBruteforce7.java
package bruteforce.KimJunYoung;

public class ProgrammersBruteforce7 {
    public static void main(String[] args) {
        Solution solved = new Solution();
        String word1 = "AAAAE";
        System.out.println(solved.solution(word1) == 6);
        solved = new Solution();
        String word2 = "I";
        System.out.println(solved.solution(word2) == 1563);
    }
    static class Solution {
        StringBuilder sb = new StringBuilder();
        boolean isDone = false;
//        int answer = 0;
        public int solution(String word) {
            char[] vowels = {'A', 'E', 'I', 'O', 'U'};
            int answer = getIndex(word, vowels, 0);
            return answer;
        }

        public int getIndex(String word, char[] vowels, int index){
            if (word.contentEquals(sb)) {
                isDone = true;
                return index;
            }
            if (sb.length() == 5) {
                return index;
            }

            for (int i = 0; i < 5; i++) {
                sb.append(vowels[i]);
                index++;
                index = getIndex(word, vowels, index);
                if (isDone) {
                    return index;
                }
                sb = new StringBuilder(sb.substring(0, sb.length() - 1));
            }
            return index;
        }
    }
}

File: brute_force/LeeJuYoung/programmers_brute_force_7.py
def solution(word):
    # 사용할 모음 리스트와 각 자리에서 나올 수 있는 단어들의 반복 패턴 길이를 정의
    vowels = ['A', 'E', 'I', 'O', 'U']
    # 자리마다의 가중치 (1자리, 2자리, 3자리 ...에서 나올 수 있는 경우의 수)
    weights = [781, 156, 31, 6, 1]  # 781 = 5^4 + 5^3 + 5^2 + 5^1 + 1
    
    result = 0  # 최종적으로 반환할 순번
    
    # 단어의 각 자리마다 해당하는 모음의 순번을 계산
    for i, char in enumerate(word):
        # 각 자리에 있는 문자가 모음 리스트에서 몇 번째인지 찾는다
        index = vowels.index(char)
        
        # 자리별 가중치를 곱해서 현재 문자가 사전에서 차지하는 순번을 더해준다
        result += index * weights[i]
        
        # 해당 문자가 사전에서 출현하는 첫 번째 단어까지의 개수를 더해줌 (자기 자신 포함)
        result += 1
    
    return result